### PR TITLE
Lazy loading Twig extensions

### PR DESCRIPTION
**Pull request by crackgm**
| Q               | A
| --------------- | -----
| Bug fix?        | no
| New feature?    | yes
| BC breaks?      | no
| Deprecations?   | no
| Related tickets | mentioned in #282 
| License         | MIT

### CODE REVIEW

File: src/Twig/Runtime/RenderBlockRuntime.php [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+        $this->blockResourceResolver = $blockResourceResolver;
+        $this->templatingEngine = $templatingEngine;
+    }
+
+    public function renderBlock(string $code, ?string $template = null): string
+    {
+        $block = $this->blockResourceResolver->findOrLog($code);
+
+        if (null !== $block) {
+            $template = $template ?? '@BitBagSyliusCmsPlugin/Shop/Block/show.html.twig';
```
Comment (senghe) [inline src/Twig/Runtime/RenderBlockRuntime.php:45]: The @BitBagSyliusCmsPlugin/Shop/Block/show.html.twig should be put into private constant

File: src/Twig/Runtime/RenderBlockRuntime.php [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+
+            return $this->templatingEngine->render($template, ['block' => $block]);
+        }
+
+        return '';
```
Comment (senghe) [inline src/Twig/Runtime/RenderBlockRuntime.php:50]: Why returning empty string? If we didn't find block, I would rather throw exception (or is it business condition?)

File: src/Twig/Runtime/RenderLinkRuntime.php [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+        ]);
+    }
+
+    public function getLinkForCode(
+        string $code,
+        array $options = []
+    ): string {
+        $page = $this->pageRepository->findOneEnabledByCode($code, $this->localeContext->getLocaleCode());
+        if (!$page) {
+            return isset($options['notFoundMessage']) ? $options['notFoundMessage'] : '';
```
Comment (senghe) [inline src/Twig/Runtime/RenderLinkRuntime.php:66]: Is this a business rule to render empty string on negative path?

File: src/Twig/Runtime/RenderMediaRuntime.php [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+
+    public function renderMedia(string $code): string
+    {
+        $media = $this->mediaResourceResolver->findOrLog($code);
+
+        if (null !== $media) {
+            return $this->mediaProviderResolver->resolveProvider($media)->render($media);
+        }
+
+        return '';
```
Comment (senghe) [inline src/Twig/Runtime/RenderMediaRuntime.php:42]: Same here about empty value

File: src/Twig/Runtime/RenderProductPagesRuntime.php [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+    {
+        $channelCode = $this->channelContext->getChannel()->getCode();
+
+        if (null !== $sectionCode) {
+            $pages = $this->pageRepository->findByProductAndSectionCode($product, $sectionCode, $channelCode);
+        } else {
+            $pages = $this->pageRepository->findByProduct($product, $channelCode);
+        }
+
+        $data = $this->sortBySections($pages);
```
Comment (senghe) [inline src/Twig/Runtime/RenderProductPagesRuntime.php:52]: I don't see any test scenario for sortings

File: src/Twig/Runtime/RenderProductPagesRuntime.php [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+        $this->pageRepository = $pageRepository;
+        $this->channelContext = $channelContext;
+        $this->templatingEngine = $templatingEngine;
+    }
+
+    public function renderProductPages(ProductInterface $product, string $sectionCode = null): string
+    {
+        $channelCode = $this->channelContext->getChannel()->getCode();
+
+        if (null !== $sectionCode) {
```
Comment (senghe) [inline src/Twig/Runtime/RenderProductPagesRuntime.php:46]: I don't see full coverage for all scenarios for this condition

File: src/Twig/Runtime/RenderMediaRuntime.php (hunk 4 above)
Comment (senghe) [inline src/Twig/Runtime/RenderMediaRuntime.php:42]: I don't see test for this return value

**Review by senghe — CHANGES_REQUESTED** [review]
1. The functionality isn't fully tested.
2. Do we have rule in entire plugin to silently return empty values when data not found?

File: src/Twig/Runtime/RenderLinkRuntime.php (hunk 3 above)
Comment (crackgm) [reply src/Twig/Runtime/RenderLinkRuntime.php:66]: It's a business rule

File: src/Twig/Runtime/RenderBlockRuntime.php (hunk 1 above)
Comment (crackgm) [reply src/Twig/Runtime/RenderBlockRuntime.php]: Extracted this string into constant

File: src/Twig/Runtime/RenderBlockRuntime.php (hunk 2 above)
Comment (crackgm) [reply src/Twig/Runtime/RenderBlockRuntime.php:52]: It's business rule

File: src/Twig/Runtime/RenderMediaRuntime.php (hunk 4 above)
Comment (crackgm) [reply src/Twig/Runtime/RenderMediaRuntime.php:42]: It's also business rule

File: src/Twig/Runtime/RenderProductPagesRuntime.php (hunk 5 above)
Comment (crackgm) [reply src/Twig/Runtime/RenderProductPagesRuntime.php]: Extracted sortBySections into new service and tested it

File: src/Twig/Runtime/RenderMediaRuntime.php (hunk 4 above)
Comment (crackgm) [reply src/Twig/Runtime/RenderMediaRuntime.php:42]: Added missing test

File: spec/Sorter/SectionsSorterSpec.php [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+    }
+
+    function it_sorts_sections(
+        PageInterface $page,
+        SectionInterface $section
+    ): void {
+        $section->getCode()->willReturn("SECTION_CODE");
+        $page->getSections()->willReturn(new ArrayCollection([$section->getWrappedObject()]));
+
+        $this->sortBySections([$page])->shouldReturn(
```
Comment (senghe) [inline spec/Sorter/SectionsSorterSpec.php:41]: When we sort, we should to write test with at least 3 elements (and good would be to have at least test with correctly sorted input and untidy one).

Comment (senghe) [reply spec/Sorter/SectionsSorterSpec.php:41]: Additionally, we have a condition inside a method - best would be to have two tests - one which goes into it and second, which doesn't.

**Review by senghe — CHANGES_REQUESTED** [review]

File: spec/Sorter/SectionsSorterSpec.php (hunk 7 above)
Comment (crackgm) [reply spec/Sorter/SectionsSorterSpec.php:41]: Added more tests

**Pull request merged**